File: src/persona_composer/factorial.py
"""Factorial (2^k) trait ablation helper."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from itertools import combinations
from pathlib import Path

from persona_composer.compose import CompositionResult, compose
from persona_composer.errors import CompositionError, ValidationError
from persona_composer.models import Module, ModuleType, SkeletonConfig
from persona_composer.parse import parse_module

DEFAULT_MAX_TRAITS = 12
_SAFE = re.compile(r"[^A-Za-z0-9._+-]+")


@dataclass
class FactorialCell:
    traits_on: list[str]
    trait_paths: list[Path]
    label: str
    result: CompositionResult | None
    error: str | None


@dataclass
class FactorialResult:
    cells: list[FactorialCell]
    timestamp: str
    trait_names: list[str]


def _as_module(
    item: Path | Module,
    *,
    module_root: Path | None,
) -> Module:
    if isinstance(item, Module):
        return item
    return parse_module(item, module_root=module_root)


def cell_label(trait_names: list[str]) -> str:
    if not trait_names:
        return "none"
    return "+".join(trait_names)


def sanitize_label(label: str) -> str:
    cleaned = _SAFE.sub("_", label).strip("._")
    return cleaned or "cell"


def _subsets_by_popcount(
    named: list[tuple[str, Path | Module]],
) -> list[list[tuple[str, Path | Module]]]:
    """Order: popcount ascending, then lex by sorted names within size."""
    ordered: list[list[tuple[str, Path | Module]]] = []
    n = len(named)
    for r in range(n + 1):
        combos = list(combinations(named, r))
        combos.sort(key=lambda c: tuple(name for name, _ in c))
        for combo in combos:
            ordered.append(list(combo))
    return ordered


def factorial_compose(
    identity: Path | Module,
    traits: list[Path | Module],
    *,
    baseline: list[Path | Module] | None = None,
    module_root: Path | None = None,
    library_root: Path | None = None,
    skeleton: SkeletonConfig | None = None,
    timestamp: str | None = None,
    max_traits: int = DEFAULT_MAX_TRAITS,
) -> FactorialResult:
    """Compose every subset of ``traits`` (2^k), with ``baseline`` always on."""
    if not traits:
        raise ValidationError("factorial requires at least one trait module")
    if len(traits) > max_traits:
        raise ValidationError(
            f"too many traits for factorial: {len(traits)} > max_traits={max_traits}"
        )

    ts = timestamp or datetime.now(timezone.utc).isoformat()
    identity_mod = _as_module(identity, module_root=module_root)
    if identity_mod.type != ModuleType.IDENTITY:
        raise ValidationError(
            f"identity must be type=identity, got {identity_mod.type.value}"
        )

    named: list[tuple[str, Path | Module]] = []
    seen_names: set[str] = set()
    for item in traits:
        mod = _as_module(item, module_root=module_root)
        if mod.type != ModuleType.TRAIT:
            raise ValidationError(
                f"factorial traits must be type=trait, got {mod.type.value} ({mod.name})"
            )
        if mod.name in seen_names:
            raise ValidationError(f"duplicate trait name in factorial list: {mod.name}")
        seen_names.add(mod.name)
        # Keep Path for trait_paths when possible
        path_or_mod: Path | Module = item if isinstance(item, Path) else mod
        named.append((mod.name, path_or_mod))

    named.sort(key=lambda t: t[0])
    trait_names = [n for n, _ in named]
    baseline_list = list(baseline or [])

    cells: list[FactorialCell] = []
    for subset in _subsets_by_popcount(named):
        names = [n for n, _ in subset]
        paths = [
            (p if isinstance(p, Path) else Path(p.path)) for _, p in subset
        ]
        label = cell_label(names)
        extras = list(baseline_list) + [item for _, item in subset]
        try:
            result = compose(
                identity_mod if isinstance(identity, Module) else identity,
                extras,
                skeleton=skeleton,
                module_root=module_root,
                library_root=library_root if library_root is not None else module_root,
                timestamp=ts,
            )
            cells.append(
                FactorialCell(
                    traits_on=names,
                    trait_paths=paths,
                    label=label,
                    result=result,
                    error=None,
                )
            )
        except CompositionError as exc:
            cells.append(
                FactorialCell(
                    traits_on=names,
                    trait_paths=paths,
                    label=label,
                    result=None,
                    error=str(exc),
                )
            )

    return FactorialResult(cells=cells, timestamp=ts, trait_names=trait_names)


def write_factorial(
    result: FactorialResult,
    out_dir: Path,
    *,
    write_prompts: bool = True,
) -> Path:
    """Write manifests/, prompts/, and index.json under ``out_dir``."""
    out_dir = Path(out_dir)
    manifests_dir = out_dir / "manifests"
    prompts_dir = out_dir / "prompts"
    manifests_dir.mkdir(parents=True, exist_ok=True)
    if write_prompts:
        prompts_dir.mkdir(parents=True, exist_ok=True)

    index_cells: list[dict] = []
    used_files: set[str] = set()

    for cell in result.cells:
        base = sanitize_label(cell.label)
        file_stem = base
        n = 2
        while file_stem in used_files:
            file_stem = f"{base}_{n}"
            n += 1
        used_files.add(file_stem)

        manifest_rel: str | None = None
        prompt_rel: str | None = None

        if cell.result is not None:
            manifest_path = manifests_dir / f"{file_stem}.json"
            manifest_path.write_text(cell.result.manifest_json(), encoding="utf-8")
            manifest_rel = f"manifests/{file_stem}.json"
            if write_prompts:
                prompt_path = prompts_dir / f"{file_stem}.xml"
                prompt_path.write_text(cell.result.prompt_xml, encoding="utf-8")
                prompt_rel = f"prompts/{file_stem}.xml"

        index_cells.append(
            {
                "label": cell.label,
                "traits_on": list(cell.traits_on),
                "manifest": manifest_rel,
                "prompt": prompt_rel,
                "error": cell.error,
            }
        )

    index = {
        "timestamp": result.timestamp,
        "trait_names": list(result.trait_names),
        "cells": index_cells,
    }
    index_path = out_dir / "index.json"
    index_path.write_text(json.dumps(index, indent=2) + "\n", encoding="utf-8")
    return index_path

File: ts/src/factorial.ts
/** Factorial (2^k) trait ablation helper. */

import { mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";

import { compose, type CompositionResult } from "./compose.js";
import { CompositionError, ValidationError } from "./errors.js";
import type { Module, SkeletonConfig } from "./models.js";
import { parseModule } from "./parse.js";

export const DEFAULT_MAX_TRAITS = 12;

export interface FactorialCell {
  traitsOn: string[];
  traitPaths: string[];
  label: string;
  result: CompositionResult | null;
  error: string | null;
}

export interface FactorialResult {
  cells: FactorialCell[];
  timestamp: string;
  traitNames: string[];
}

function asModule(
  item: string | Module,
  moduleRoot?: string,
): Module {
  if (typeof item !== "string") return item;
  return parseModule(item, { moduleRoot });
}

export function cellLabel(traitNames: string[]): string {
  if (!traitNames.length) return "none";
  return traitNames.join("+");
}

export function sanitizeLabel(label: string): string {
  const cleaned = label.replace(/[^A-Za-z0-9._+-]+/g, "_").replace(/^[._]+|[._]+$/g, "");
  return cleaned || "cell";
}

function combinations<T>(items: T[], r: number): T[][] {
  if (r === 0) return [[]];
  if (r > items.length) return [];
  const out: T[][] = [];
  const go = (start: number, acc: T[]) => {
    if (acc.length === r) {
      out.push([...acc]);
      return;
    }
    for (let i = start; i < items.length; i++) {
      acc.push(items[i]!);
      go(i + 1, acc);
      acc.pop();
    }
  };
  go(0, []);
  return out;
}

function subsetsByPopcount<T extends { name: string }>(
  named: T[],
): T[][] {
  const ordered: T[][] = [];
  for (let r = 0; r <= named.length; r++) {
    const combos = combinations(named, r);
    combos.sort((a, b) => {
      const aa = a.map((x) => x.name).join("\0");
      const bb = b.map((x) => x.name).join("\0");
      return aa < bb ? -1 : aa > bb ? 1 : 0;
    });
    ordered.push(...combos);
  }
  return ordered;
}

export interface FactorialComposeOptions {
  baseline?: Array<string | Module>;
  moduleRoot?: string;
  libraryRoot?: string;
  skeleton?: SkeletonConfig;
  timestamp?: string;
  maxTraits?: number;
}

export function factorialCompose(
  identity: string | Module,
  traits: Array<string | Module>,
  options: FactorialComposeOptions = {},
): FactorialResult {
  const maxTraits = options.maxTraits ?? DEFAULT_MAX_TRAITS;
  if (!traits.length) {
    throw new ValidationError("factorial requires at least one trait module");
  }
  if (traits.length > maxTraits) {
    throw new ValidationError(
      `too many traits for factorial: ${traits.length} > max_traits=${maxTraits}`,
    );
  }

  const ts = options.timestamp ?? new Date().toISOString();
  const identityMod = asModule(identity, options.moduleRoot);
  if (identityMod.type !== "identity") {
    throw new ValidationError(
      `identity must be type=identity, got ${identityMod.type}`,
    );
  }

  type Named = { name: string; item: string | Module; path: string };
  const named: Named[] = [];
  const seen = new Set<string>();
  for (const item of traits) {
    const mod = asModule(item, options.moduleRoot);
    if (mod.type !== "trait") {
      throw new ValidationError(
        `factorial traits must be type=trait, got ${mod.type} (${mod.name})`,
      );
    }
    if (seen.has(mod.name)) {
      throw new ValidationError(
        `duplicate trait name in factorial list: ${mod.name}`,
      );
    }
    seen.add(mod.name);
    named.push({
      name: mod.name,
      item,
      path: typeof item === "string" ? item : mod.path,
    });
  }
  named.sort((a, b) => a.name.localeCompare(b.name));
  const traitNames = named.map((n) => n.name);
  const baseline = options.baseline ?? [];

  const cells: FactorialCell[] = [];
  for (const subset of subsetsByPopcount(named)) {
    const names = subset.map((s) => s.name);
    const paths = subset.map((s) => s.path);
    const label = cellLabel(names);
    const extras = [...baseline, ...subset.map((s) => s.item)];
    try {
      const result = compose(identity, extras, {
        skeleton: options.skeleton,
        moduleRoot: options.moduleRoot,
        libraryRoot: options.libraryRoot ?? options.moduleRoot,
        timestamp: ts,
      });
      cells.push({
        traitsOn: names,
        traitPaths: paths,
        label,
        result,
        error: null,
      });
    } catch (exc) {
      if (exc instanceof CompositionError) {
        cells.push({
          traitsOn: names,
          traitPaths: paths,
          label,
          result: null,
          error: exc.message,
        });
      } else {
        throw exc;
      }
    }
  }

  return { cells, timestamp: ts, traitNames };
}

export function writeFactorial(
  result: FactorialResult,
  outDir: string,
  options: { writePrompts?: boolean } = {},
): string {
  const writePrompts = options.writePrompts !== false;
  const manifestsDir = path.join(outDir, "manifests");
  const promptsDir = path.join(outDir, "prompts");
  mkdirSync(manifestsDir, { recursive: true });
  if (writePrompts) mkdirSync(promptsDir, { recursive: true });

  const indexCells: Record<string, unknown>[] = [];
  const used = new Set<string>();

  for (const cell of result.cells) {
    const base = sanitizeLabel(cell.label);
    let fileStem = base;
    let n = 2;
    while (used.has(fileStem)) {
      fileStem = `${base}_${n}`;
      n += 1;
    }
    used.add(fileStem);

    let manifestRel: string | null = null;
    let promptRel: string | null = null;

    if (cell.result) {
      const manifestPath = path.join(manifestsDir, `${fileStem}.json`);
      writeFileSync(manifestPath, cell.result.manifestJson(), "utf-8");
      manifestRel = `manifests/${fileStem}.json`;
      if (writePrompts) {
        const promptPath = path.join(promptsDir, `${fileStem}.xml`);
        writeFileSync(promptPath, cell.result.promptXml, "utf-8");
        promptRel = `prompts/${fileStem}.xml`;
      }
    }

    indexCells.push({
      label: cell.label,
      traits_on: cell.traitsOn,
      manifest: manifestRel,
      prompt: promptRel,
      error: cell.error,
    });
  }

  const indexPath = path.join(outDir, "index.json");
  writeFileSync(
    indexPath,
    JSON.stringify(
      {
        timestamp: result.timestamp,
        trait_names: result.traitNames,
        cells: indexCells,
      },
      null,
      2,
    ) + "\n",
    "utf-8",
  );
  return indexPath;
}
